File: meme_inator_front/lib/features/profiles/domain/interfaces/iprofile_repository.dart
// lib/features/profiles/domain/interfaces/iprofile_repository.dart
import 'package:meme_inator_front/core/results.dart';
import 'package:meme_inator_front/features/profiles/domain/entities/profile_entity.dart';

abstract class IProfileRepository {
  // Public profile
  Future<Result<ProfileEntity>> getPublicProfile({
    required String username,
    String? viewerUserId,
  });

  /// Fetch a public profile including whether the authenticated user follows it.
  /// [profileOwnerUserId] is the UUID of the profile being viewed.
  /// The viewer's ID is taken from the backend's auth context automatically.
  Future<Result<ProfileEntity>> getPublicProfileWithFollowshipContext({
    required String profileOwnerUserId,
    String? viewerUserId, // optional, backend extracts from token anyway
    List<String>? fields,
  });

  // Own profile
  Future<Result<ProfileEntity>> getMyProfile();
  Future<Result<ProfileEntity>> createProfile();
  Future<Result<ProfileEntity>> patchMyProfile(
    Map<String, dynamic> partialData,
  );
  Future<Result<ProfileEntity>> replaceMyProfile(Map<String, dynamic> fullData);

  // Counters
  Future<Result<ProfileEntity>> updateCounters(Map<String, int> increments);

  // Media sync
  Future<Result<ProfileEntity>> syncMedia(Map<String, dynamic> mediaPayload);

  // Posts - returns raw map that can be converted by usecase
  Future<Result<Map<String, dynamic>>> getProfilePosts({
    required String username,
    String? cursor,
    int pageSize = 10,
  });

  // Image URLs (presigned)
  Future<Result<Map<String, String>>> getProfileImageUrls();

  // Validation (optional – could be used before sending)
  Future<Result<void>> validateProfile(Map<String, dynamic> profileData);
}

File: meme_inator_back/apps/profiles/domain/irepositories/i_profile_repository.py
from typing import Optional, Dict, Any, List, Union
from uuid import UUID
from apps.profiles.domain.entities.profile_light_entity import ProfileLightEntity
from core.results import Result
from apps.profiles.domain.entities.profile_entity import ProfileEntity

class IProfileRepository:
    """Interface for profile repository operations."""

    def get_public_profile(
            self, 
            username: str, 
            viewer_user_id: Optional[UUID], 
            fields: Optional[List[str]] = None
        ) -> Result[ProfileEntity|ProfileLightEntity]:
        """Fetch a public profile by username."""
        ...

    def get_my_profile(
            self, 
            user_id: UUID, 
            viewer_user_id: Optional[UUID]
        ) -> Result[ProfileEntity|ProfileLightEntity]:
        """Fetch the profile of the currently authenticated user."""
        ...

    def create_profile(self, user_id: UUID) -> Result[ProfileEntity]:
        """Create a default profile for a user."""
        ...

    def patch_my_profile(self, user_id: UUID, partial_data: Dict[str, Any]) -> Result[ProfileEntity]:
        """Partially update the authenticated user's profile."""
        ...

    def replace_my_profile(self, user_id: UUID, full_data: Dict[str, Any]) -> Result[ProfileEntity]:
        """Fully replace the authenticated user's profile."""
        ...

    def update_counters(self, user_id: UUID, increments: Dict[str, int]) -> Result[ProfileEntity]:
        """Atomically update numeric counters (e.g., followers_count)."""
        ...

    def sync_media(self, user_id: UUID, media_payload: Dict[str, Any]) -> Result[ProfileEntity]:
        """Update media URLs after upload (profile pic, header, etc.)."""
        ...

    def get_profile_posts(self, username: str, cursor: Optional[str] = None, page_size: int = 10) -> Result[Dict[str, Any]]:
        """Fetch paginated posts for a profile."""
        ...

    def get_profile_image_urls(self, user_id: UUID) -> Result[Dict[str, str]]:
        """Get presigned or public URLs for profile images."""
        ...

    def get_user_id_by_username(self, username: str) -> Result[UUID]:
        """Return the user ID for a given username."""
        ...
    
    def get_public_profile_by_user_id(
        self, 
        user_id: UUID, 
        viewer_user_id: Optional[UUID], 
        fields: Optional[List[str]] = None
    ) -> Result[Union[ProfileEntity, ProfileLightEntity]]:
        ...